### PR TITLE
feat(sourcemaps): Check for release or debug ids before upload

### PR DESCRIPTION
**Pull request by loewenheim**
This should take care of #1547.

Unfortunately, it requires a significant refactoring of `SourceMapProcessor`. Parts of `add_sourcemap_references` and `inject_debug_ids` have been extracted into a new method `collect_sourcemap_references` that collects the sourcemap files referenced from minified sources into a map saved in the processor. This method is called from `flush_pending_sources`. There is still some duplication of the logic in `unpack_ram_bundle` which could probably be subsumed into this, but I'm not confident I can do it correctly.

I decided to bail with an error if no files have debug ids and use `warn!` if only some of them do. I think these choices and the error messages are still up for debate, though.

Also: any idea how we could test this?

### CODE REVIEW

**Comment by iambriccardo** [comment]
I would test it in the following way:
1. Create a new bundle
2. Run the upload command with the new flag
3. Expect an error
--
1. Create a new bundle
2. Run the inject command
3. Manually remove debug ids from one or more files
4. Run the upload command with the new flag
5. Expect a warning
--
1. Create a new bundle
3. Run the upload command with the new flag and release
4. Expect no error

File: src/utils/sourcemaps.rs [hunk 1]
```diff
@@ -650,6 +677,40 @@ impl SourceMapProcessor {
...
+                if self.debug_ids.contains_key(source_url) {
+                    files_with_debug_id = true;
+                } else {
+                    files_without_debug_id.insert(source_url.clone());
+                }
+            }
+
+            // No debug ids on any files -> can't upload
+            if !files_without_debug_id.is_empty() && !files_with_debug_id {
+                bail!("Cannot upload: No release or debug ids provided");
```
Comment (iambriccardo) [inline src/utils/sourcemaps.rs:701]:
```suggestion
                bail!("Cannot upload: You must either specify a release or have debug ids injected into your sources");
```

**Review by iambriccardo — APPROVED** [review]
LGTM

**Comment by loewenheim** [comment]
Locally running this has raised a question: what should happen if a minified file points to a source file that can't be found on the local file system? Should that raise a warning for a missing debug id?

**Comment by iambriccardo** [comment]
> Locally running this has raised a question: what should happen if a minified file points to a source file that can't be found on the local file system? Should that raise a warning for a missing debug id?

If you mean that it points to an invalid location for the sourcemap, I would also consider this as an "injection" failure and we could warn the user.

**Comment by kamilogorek** [comment]
I know those are fixtures, but what's inside them is mostly irrelevant to us. Can we make them smaller than 50k LoC? 😅

**Pull request merged**